src/main/scala/nodes/learning/LBFGS.scala

```diff
@@ -0,0 +1,242 @@
+      regParam: Double,
+      weightsIncludeBias: Boolean = false): DenseMatrix[Double] = {
+
+    val lossHistory = mutable.ArrayBuilder.make[Double]
+    val numExamples = data.count
+    val numFeatures = data.first.length
+    val numClasses = labels.first.length
+
+    val costFun = new CostFun(data, labels, gradient, regParam, numExamples, numFeatures,
+      numClasses)
+
+    val lbfgs = new BreezeLBFGS[DenseVector[Double]](maxNumIterations, numCorrections, convergenceTol)
+
+    val initialWeights = DenseVector.zeros[Double](numFeatures * numClasses)
+
+    val states =
+      lbfgs.iterations(new CachedDiffFunction(costFun), initialWeights)
+
+    /**
+      * NOTE: lossSum and loss is computed using the weights from the previous iteration
+      * and regVal is the regularization value computed in the previous iteration as well.
+      */
+    var state = states.next()
+    while (states.hasNext) {
+      lossHistory += state.value
+      state = states.next()
+    }
+    lossHistory += state.value
+    val finalWeights = state.x.asDenseMatrix.reshape(numFeatures, numClasses)
+
+    val lossHistoryArray = lossHistory.result()
+
+    logInfo("LBFGS.runLBFGS finished. Last 10 losses %s".format(
+      lossHistoryArray.takeRight(10).mkString(", ")))
+
+    finalWeights
+  }
+
+  /**
+    * CostFun implements Breeze's DiffFunction[T], which returns the loss and gradient
+    * at a particular point (weights). It's used in Breeze's convex optimization routines.
+    */
+  private class CostFun[T <: Vector[Double]](
+      data: RDD[T],
+      labels: RDD[DenseVector[Double]],
+      gradient: Gradient[T],
+      regParam: Double,
+      numExamples: Long,
+      numFeatures: Int,
+      numClasses: Int,
+      weightsIncludeBias: Boolean = false) extends DiffFunction[DenseVector[Double]] {
+
+    override def calculate(weights: DenseVector[Double]): (Double, DenseVector[Double]) = {
+      val weightsMat = weights.asDenseMatrix.reshape(numFeatures, numClasses)
+      // Have a local copy to avoid the serialization of CostFun object which is not serializable.
+      val bcW = data.context.broadcast(weightsMat)
+      val localGradient = gradient
+      val localNumFeatures = numFeatures
+      val localNumClasses = numClasses
+
+      val partitionGradientLosses = data.zipPartitions(labels) {
+        (partitionFeatures, partitionLabels) =>
+          Iterator.single(localGradient.compute(localNumFeatures, localNumClasses, partitionFeatures, partitionLabels, bcW.value))
+      }
+
+      val (gradientSum, lossSum) = MLMatrixUtils.treeReduce(
+        partitionGradientLosses,
+        (a: (DenseMatrix[Double], Double), b: (DenseMatrix[Double], Double)) => {
+          a._1 += b._1
+          (a._1, a._2 + b._2)
+        }
+      )
+
+      // total loss = lossSum / nTrain + 1/2 * lambda * norm(W)^2
+      val normWSquared = if (weightsIncludeBias) {
+        math.pow(norm(weights(0 until (weights.length - 1))), 2)
```

Comment:
Isn't weights a matrix of size `d x k` In that case we'll need to exclude all the bias terms per-each class rather than just one bias term ?